### PR TITLE
[PAPP-33470] Bugfix: Recurrent multipart parsing

### PR DESCRIPTION
**Pull request by mposluszny-splunk**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [x] The manual_readme_content.md has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [ ] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [ ] Verify all checks are passing.
- [x] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [x] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe):

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers),
  please document them in the `manual_readme_content.md`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `manual_readme_content.md` all methods (eg, OAuth) used to authenticate
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `manual_readme_content.md`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

**Review by bbielinski-splunk — APPROVED** [review]
Looks good, I have left some suggestions to address.

File: gsgmail.json [hunk 1]
```diff
@@ -954,6 +954,12 @@
                     "data_type": "boolean",
                     "default": false,
                     "order": 2
+                },
+                "extract_nested": {
+                    "description": "Works when `extract_attachments` is set to `true`. Extracts attachments from nested email attachments.",
+                    "data_type": "boolean",
+                    "default": false,
```
[the remark this answers is not in the document]
Comment (mposluszny-splunk) [reply gsgmail.json:961]: Well, default behavior was broken. Cause on gmail page I cannot see attachments of attachments and customer do not expect to see attachments that are part of other attachments :D. That's why I left it on false

File: gsgmail_connector.py [hunk 2]
```diff
@@ -407,6 +338,134 @@ def _handle_run_query(self, param):
...
+        extract_attachments=False,
+        extract_nested=False,
+    ):
+        self._init_detail_fields(email_details)
+
+        def traverse(part, is_attachment=False):
+            if not is_attachment:
+                self._parse_email_details(part, email_details)
+
+            ret_val = phantom.APP_SUCCESS
```
[the remark this answers is not in the document]
Comment (mposluszny-splunk) [reply gsgmail_connector.py:448]: Good point, I'll do a rename

File: gsgmail.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (bbielinski-splunk) [reply gsgmail.json:961]: Ok, if default behavior was broken then let's leave it on false

File: gsgmail_connector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mposluszny-splunk) [reply gsgmail_connector.py:448]: bbielienski, I've digged dipper into this RetVal2 thing, it's basically the same class as RetVal. It's just a wrapper that states 2 element tupl will be returned. More over ph_status is reserved for import name so I cannot use that. I'll stay with ret_val for now. But we should address in case of bigger refactors

File: gsgmail_connector.py [hunk 3]
```diff
@@ -910,7 +978,7 @@ def handle_action(self, param):
             in_json['user_session_token'] = session_id
             connector._set_csrf_info(csrftoken, headers['Referer'])
 
-        ret_val = connector._handle_action(json.dumps(in_json), None)
-        print(json.dumps(json.loads(ret_val), indent=4))
+        ph_status = connector._handle_action(json.dumps(in_json), None)
```
Comment (bbielinski-splunk) [inline gsgmail_connector.py:981]: You have mentioned that 'ph_status' is reserved for imports. Is this change intended then?

**Review by bbielinski-splunk — APPROVED** [review]

File: gsgmail_connector.py (hunk 3 above)
Comment (mposluszny-splunk) [reply gsgmail_connector.py:981]: completely not :D

File: gsgmail_connector.py [hunk 4]
```diff
@@ -407,6 +338,137 @@ def _handle_run_query(self, param):
...
+            },
+            "run_automation": False,
+            "source_data_identifier": None,
+        }
+
+    def _parse_email_details(self, part, email_details):
+        headers = self._get_email_headers_from_part(part)
+        # split out important headers (for output table rendering)
+        if headers.get("to"):
+            email_details["to"] = headers.get("to")
```
Comment (bb-splunk) [inline gsgmail_connector.py:368]: No need for `get` at this point

File: gsgmail_connector.py [hunk 5]
```diff
@@ -407,6 +338,137 @@ def _handle_run_query(self, param):
 
         return action_result.set_status(phantom.APP_SUCCESS)
 
+    def _body_from_part(self, part):
+        charset = part.get_content_charset() or "utf8"
```
Comment (bb-splunk) [inline gsgmail_connector.py:342]: Isn't it 'utf-8'?

File: gsgmail_connector.py [hunk 6]
```diff
@@ -407,6 +338,137 @@ def _handle_run_query(self, param):
...
+        # decode the base64 unicode bytestring into plain text
+        plain_body = part.get_payload(decode=True).decode(
+            encoding=charset, errors="ignore"
+        )
+        # Add to list of plan text bodies
```
Comment (bb-splunk) [inline gsgmail_connector.py:347]: 1) plaIn
2) This comment doesn't really make sense nor is really necessary

**Review by bb-splunk — CHANGES_REQUESTED** [review]

File: gsgmail_connector.py (hunk 5 above)
Comment (mposluszny-splunk) [reply gsgmail_connector.py]: It is, checked with emai lib charsets. All of that are left overs from previous implementation

File: gsgmail_connector.py [hunk 7]
```diff
@@ -407,6 +338,137 @@ def _handle_run_query(self, param):
...
+        self,
+        action_result,
+        msg,
+        email_details,
+        extract_attachments=False,
+        extract_nested=False,
+    ):
+        self._init_detail_fields(email_details)
+
+        def traverse(part, in_attachment=False):
```
[the remark this answers is not in the document]
Comment (mposluszny-splunk) [reply gsgmail_connector.py]: No reason, leftover from prototyping recurrency based solution

File: gsgmail_connector.py [hunk 8]
```diff
@@ -407,6 +338,148 @@ def _handle_run_query(self, param):
...
+    @staticmethod
+    def _is_attachment(part):
+        return "attachment" in str(part.get("Content-Disposition"))
+
+    def _init_detail_fields(self, email_details):
+        email_details["plain_bodies"] = []
+        email_details["html_bodies"] = []
+        email_details["email_headers"] = []
+
+    def _parse_email_bodies(self, email_details):
```
Comment (bb-splunk) [inline gsgmail_connector.py:423]: `format`/`join` would better describe what this does

**Review by bb-splunk — APPROVED** [review]

File: gsgmail_connector.py (hunk 8 above)
Comment (mposluszny-splunk) [reply gsgmail_connector.py:423]: agreed

**Review by ivellios — APPROVED** [review]
